File: planners/downward_gpu/src/search/heuristics/pog2.cc
#include "pog2.h"

#include <chrono>


PredicateObjectGraph2::PredicateObjectGraph2(const std::string& domain_file, const std::string& problem_file) : BaseGraph(domain_file, problem_file) {
    // Graph
    gen_nodes();
    gen_init(goals, "goal");
}        

void PredicateObjectGraph2::gen_nodes() {
    // Create a set of all node types from constants and objects
    std::set<std::string> all_node_types;
    for (const auto& pair : constants) {
	all_node_types.insert(pair.first);
    }
    for (const auto& pair : objects) {
	all_node_types.insert(pair.first);
    }

    // Iterate over sorted node types
    for (const auto& key : all_node_types) {
	if (constants.find(key) != constants.end()) {
            for (const auto &obj : constants[key]) {
		nodes_index[obj] = nodes.size() / 2;
                nodes.emplace_back(0);     // unachievd,
		nodes.emplace_back(node_classes[key]);
            }
	}
	if (objects.find(key) != objects.end()) {
            for (const auto &obj : objects[key]) {
		nodes_index[obj] = nodes.size() / 2;
		nodes.emplace_back(0);     // unachievd,
		nodes.emplace_back(node_classes[key]);
            }
        }
    }

    // Generate nodes for predicate types
    for (const auto &predicate : predicates) {
	// Since we don't create predicate nodes, we use their class index instead.
        nodes_index[predicate] = node_classes[predicate];
    }
}


void PredicateObjectGraph2::gen_init(const std::set<std::string>& predicates_strs, const std::string& node_type) {
    for (const auto& predicate_str : predicates_strs) {
	std::istringstream iss(predicate_str);
	std::string predicate;
	std::vector<std::string> items;
        std::string item;
        
	while (iss >> item) {
	    items.push_back(item);
	}

	if (items.empty()) continue;
	int param_nums = items.size();

	// Add fact (goal / state) node
        int fact_index = nodes.size() / 2;
        if (node_type == "goal"){
	    nodes.emplace_back(1);   // unachieved
	    nodes_goal_index[predicate_str] = fact_index;  // goal
	} else {
	    nodes.emplace_back(0);   // current state             
        }

	nodes.emplace_back(nodes_index[items[0]]);
        for (int i = 1; i < param_nums; i++) {
            if (items[i].empty())
		continue;
	    // Add edges, predicate to fact
            edges_1.emplace_back(fact_index);
            edges_2.emplace_back(nodes_index[items[i]]);
            edges_1.emplace_back(nodes_index[items[i]]);
            edges_2.emplace_back(fact_index);
            edge_attrs.push_back(i - 1);
	    edge_attrs.push_back(i - 1);            
        }
    }
}

std::tuple<std::vector<int>, std::vector<int>, std::vector<int>, std::vector<int>> PredicateObjectGraph2::gen_nodes_edges(const std::vector<std::pair<std::string, std::vector<int>>>& states) {
    // Use the copy constructor
    int fact_index = nodes.size() / 2;
    std::vector<int> new_nodes(nodes);
    std::vector<int> new_edges_1(edges_1);
    std::vector<int> new_edges_2(edges_2);    
    std::vector<int> new_edge_attrs(edge_attrs);

    for (const auto &state : states) {
	const std::string& predicate_str = state.first;
	const std::vector<int>& items = state.second;
        int param_nums = items.size();

	if (predicate_str.empty()) continue;

	if (nodes_goal_index.find(predicate_str) != nodes_index.end()) {
	    int index = nodes_goal_index.at(predicate_str) * 2; // * 2 changes to the original length
	    new_nodes[index] = 2;
	    continue;
	}

        new_nodes.emplace_back(0);            // unachievd,
	new_nodes.emplace_back(items[0]);     // current state, 
        for (int i = 1; i < param_nums; i++) {
            // int node_id = nodes_index.at(items[i]);
	    int node_id = items[i];
            new_edges_1.emplace_back(fact_index);
	    new_edges_2.emplace_back(node_id);
            new_edges_1.emplace_back(node_id);		
            new_edges_2.emplace_back(fact_index);
	    new_edge_attrs.emplace_back(i - 1);
            new_edge_attrs.emplace_back(i - 1);
	}
	fact_index+=1;
    }

    return {new_nodes, new_edges_1, new_edges_2, new_edge_attrs};
}

File: representation/pog2.py
import copy
import logging

import torch
from .base_graph import BaseGraph

logger = logging.getLogger('Logger')


class PredicateObjectGraph2(BaseGraph):

    def __init__(self, domain_pddl, problem_pddl):
        super().__init__(domain_pddl, problem_pddl)

    def gen_nodes(self):
        nodes = []
        nodes_index = {}
        node_names = []
        node_colors = []

        # gen objects
        all_node_types = set(list(self.constants.keys()) + list(self.objects.keys()))
        for key in sorted(all_node_types):
            # add constant nodes
            if key in self.constants:
                for obj in self.constants[key]:
                    nodes.append([0, self.node_class.index(key)])
                    node_names.append(obj)
                    node_colors.append(0)
                    nodes_index[obj] = len(nodes_index)

            if key in self.objects:
                for obj in self.objects[key]:
                    nodes.append([0, self.node_class.index(key)])
                    node_names.append(obj)
                    node_colors.append(0)
                    nodes_index[obj] = len(nodes_index)

        return nodes, nodes_index, node_names, node_colors

    def gen_init(self, predicates, node_type, debug=False):
        # gen facts
        for predicate_str in predicates:
            items = predicate_str.split(" ")
            param_nums = len(items) - 1
            predicate = items[0]
            predicate_str = predicate_str.replace(" ", "_")

            if debug:
                if predicate not in self.predicates:
                    print(f"ERROR: gen_goal_edges, {predicate} not in predicates {self.predicates}")

                if len(self.predicates[predicate]) != param_nums:
                    print(f"ERROR: The number of params is not right!!! {len(self.predicates[predicate])} {param_nums}")

            # add fact (goal / state) node
            # self.nodes.append([0, self.node_class.index(node_type)])  #  goal or init
            if node_type == "goal":
                self.nodes.append([1, self.node_class.index(predicate)])
                self.node_names.append(predicate)
                self.node_colors.append(2)
            else:
                self.nodes.append([0, self.node_class.index(predicate)])
                self.node_names.append(predicate)
                self.node_colors.append(1)
            self.nodes_index[predicate_str] = len(self.nodes_index)
            fact_index = len(self.nodes_index) - 1

            for i in range(param_nums):
                # fact to obj
                if items[i + 1] == "" or items[i + 1] == " ":
                    continue

                self.edges.append([fact_index, self.nodes_index[items[i + 1]]])
                self.edges.append([self.nodes_index[items[i + 1]], fact_index])
                self.edge_attrs.append(i)
                self.edge_attrs.append(i)

            # self.x = torch.cat(self.nodes, dim=0)
            self.x = torch.tensor(self.nodes)

    # Make sure this function invoking is after gen_nodes_edges
    def get_node_names_colors(self, predicates):
        node_names = copy.copy(self.node_names)
        node_colors = copy.copy(self.node_colors)
        for (predicate, items) in predicates:
            if predicate in ('', ' '):
                continue

            param_nums = len(items)
            predicate_str = predicate + "_" + "_".join(items) if param_nums > 0 else predicate
            if predicate_str in self.nodes_index:
                index = self.nodes_index[predicate_str]
                node_names[index] = predicate
                node_colors[index] = 3
                continue

            node_names.append(predicate)
            node_colors.append(1)

        return node_names, node_colors

    def gen_nodes_edges(self, predicates, node_type, debug=False):
        # gen facts
        nodes = self.x.clone()
        node_num = len(self.nodes_index)
        new_nodes = []
        new_edges = []  # [[] for i in range(self.largest_predicate + 1)]
        new_edge_attrs = []

        for (predicate, items) in predicates:
            # items = predicate_str.split(" ")
            # predicate = items[0]
            if predicate in ('', ' '):
                continue

            param_nums = len(items)
            predicate_str = predicate + "_" + "_".join(items) if param_nums > 0 else predicate
            # predicate_index.append(predicate_str)
            # print(predicate_str)
            if node_type == "state" and predicate_str in self.nodes_index:
                index = self.nodes_index[predicate_str]
                nodes[index, 0] = 2
                continue

            if debug:
                if predicate not in self.predicates:
                    print(f"ERROR: gen_goal_edges, {predicate} not in predicates {self.predicates}")

                if len(self.predicates[predicate]) != param_nums:
                    print(f"ERROR: The number of params is not right!!! {len(self.predicates[predicate])} {param_nums}")

            # add fact (goal / state) node
            new_nodes.append([0, self.node_class.index(predicate)])
            # new_nodes.append(self._one_hot_node(self.node_class.index(node_type), len(self.node_class)).unsqueeze(0))
            node_num += 1
            fact_index = node_num - 1

            for i in range(param_nums):
                # fact to obj
                new_edges.append([fact_index, self.nodes_index[items[i]]])
                new_edges.append([self.nodes_index[items[i]], fact_index])
                new_edge_attrs.append(i)
                new_edge_attrs.append(i)

        # edges = []
        # for i in range(self.largest_predicate + 1):
        #     edges.append(torch.cat((torch.tensor(self.edges[i], dtype=torch.long), torch.tensor(new_edges[i], dtype=torch.long)), dim=0).t().contiguous())

        nodes = torch.cat((nodes, torch.tensor(new_nodes)), dim=0)
        edges = torch.cat((torch.tensor(self.edges, dtype=torch.long), torch.tensor(new_edges, dtype=torch.long)), dim=0).t()
        new_edge_attrs = torch.cat((torch.tensor(self.edge_attrs, dtype=torch.long), torch.tensor(new_edge_attrs, dtype=torch.long)), dim=0)

        return nodes, edges, new_edge_attrs
